jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/DefaultServlet.java

```diff
@@ -487,44 +473,70 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
                 }
 
                 // no content
-                resp.sendError(404);
+                httpServletResponse.sendError(404);
             }
             else
             {
-                ServletCoreRequest coreRequest = new ServletCoreRequest(req);
-                ServletCoreResponse coreResponse = new ServletCoreResponse(coreRequest, resp, included);
-
+                // lookup the core request and response as wrapped by the ServletContextHandler
+                ServletContextRequest servletContextRequest = ServletContextRequest.getServletContextRequest(httpServletRequest);
+                ServletContextResponse servletContextResponse = servletContextRequest.getServletContextResponse();
+                ServletChannel servletChannel = servletContextRequest.getServletChannel();
+
+                // If the servlet request has not been wrapped,
+                // we can use the core request directly,
+                // otherwise wrap the servlet request as a core request
+                Request coreRequest = httpServletRequest instanceof ServletApiRequest
+                    ? servletChannel.getRequest()
+                    : new ServletCoreRequest(httpServletRequest);
+
+                // If the servlet response has not been wrapped and has not been written to,
+                // we can use the core response directly,
+                // otherwise wrap the servlet response as a core response
+                Response coreResponse = httpServletResponse instanceof ServletApiResponse && !servletContextResponse.isWritingOrStreaming()
+                    ? servletChannel.getResponse()
+                    : new ServletCoreResponse(coreRequest, httpServletResponse, included);
+
+                // If the core response is already committed then do nothing more
                 if (coreResponse.isCommitted())
                 {
                     if (LOG.isDebugEnabled())
                         LOG.debug("Response already committed for {}", coreRequest.getHttpURI());
                     return;
                 }
 
+                // Get the content length before we may wrap the content
+                long contentLength = content.getContentLengthValue();
+
                 // Servlet Filters could be interacting with the Response already.
-                if (coreResponse.isHttpServletResponseWrapped() ||
-                    coreResponse.isWritingOrStreaming())
-                {
+                if (coreResponse instanceof  ServletCoreResponse)
```

Comment:
Nit: there's an extra space between `instanceof` and `ServletCoreResponse`; I wonder how that's not breaking checkstyle.